Introduce an "open source notes and installation guide" document (#17)
EnterpriseDB/tpa: docs/src/opensourcetpa.md

# Open source TPA

> Trusted Postgres Architect has historically been built to address the deployment needs of EDB customers. Among its playbooks is the ability to deploy EDB-recommended architectures using EDB repositories containing EDB versions of Postgres and other tools.

Reviewer: playbook is an Ansible-specific thing and there are no playbooks in TPA

Author: Will edit that out.

Author: Reworked

As an open source deployment tool, TPA can perform similar tasks using open source repositories. This section of the documentation covers what an open source user of TPA needs to know: How to set up and configure TPA and deploy a basic configuration of open source Postgres tools.

## Setting up TPA

### Prerequisites

Before you install TPA, you must install the required packages: Python 3, git, openvpn and patch.

#### Debian

```
sudo apt-get install python3.7 python3-pip python3-venv git openvpn patch
```

> #### Ubuntu

Reviewer: I think for both Debian and Ubuntu we can just do `sudo apt-get install python3 python3-pip python3-venv git openvpn patch` there's no need to specify a Python version as all supported deb distros have a Python that's new enough these days.

Author: done

```
sudo apt-get install python3.6 python3-pip python3-venv git openvpn patch

```

#### Redhat, Rocky or AlmaLinux (RHEL7)

```
sudo yum install python3 python3-pip epel-release git openvpn patch
```

#### Redhat, Rocky or AlmaLinux (RHEL8)

```
sudo yum install python36 python3-pip epel-release git openvpn patch
```

> #### macOS

Reviewer: We don't support macOS so it's questionable whether this should be here at all (same for the 'install from source' page). At minimum this should have a warning.

Author: Removed from here.

```
git clone https://github.com/enterprisedb/tpa.git
```

This creates a `tpa` directory in the current directory.

You can now create a working tpa environment from your current directory by running:

```
./tpa/bin/tpaexec setup
```

This will create the Python virtual environment that TPA will use in future. All needed packages are installed in this environment. To test this configured correctly, run the following:

```
./tpa/bin/tpaexec selftest
```

At this point we recommend you add the `tpa/bin` directory to your path with:

```
export PATH=$PATH:$HOME/tpa/bin/
```

This is assuming you have installed tpa in your home directory. Modify the setting as appropriate if you have installed it in another location.

Add this export to your `.bashrc` or other shell profile file to ensure the tpaexec command is always available.

## Configuring a deployment on Docker

You are now ready to configure a deployment using TPA. For the example here, we will run TPA on an Ubuntu system, but the considerations are similar for most Linux systems.

We are going to deploy the example deployment onto Docker, so before we do that, we need to install Docker-CE.
### Installing Docker

> On Ubuntu, install Docker by running:

Reviewer: Why instructions only for Ubuntu? Are these official? The Docker website seems to have much longer instructions involving adding repos etc, which is what I've always used.

Author: Experience doing the quickstarts is that the repo CE build is far more reliable.

Author: But will add a "For other..." line going to the Docker docs. (If we brought over any of that content it would mean having to keep it in sync). Also changed the reference to Ubuntu to Debian/Ubuntu.

Run this command, and apparently, nothing will happen on the command line. But you will find a directory called `demo` has been created containing some files including a `config.yml` file which is a blueprint for our new deployment.

## Provisioning the deployment

Now we are ready to create the containers (or virtual machines) on which we will run our new deployment. This can be achieved with the `provision` command. Run:

```
tpaexec provision demo
```

You will see TPA work through the various operations needed to prepare for deployment of your configuration. 

## Deploying

Once provisioned, you can move on to deployment. The installs, if needed, operating systems and system packages. It then installs the requested Postgres architecture and performs all the needed configuration. 

```
tpaexec deploy demo
```

You will see TPA work through the various operations needed to deploy your configuration.

## Testing

You can quickly test your newly deployed configuration using the tpaexec `test` command which will run pgbench on your new database.

```
tpaexec test demo
```

## Connecting

> To get to a psql prompt, the simplest route is to log into one of the containers (or VMs or host depending on configuration) using SSH. Run 

Reviewer: I find the simplest route is `docker exec -it hostname /bin/bash` any reason to use SSH beyond documenting these potentially useful commands?

Reviewer: We're using Docker in this guide though, but I see your point. Perhaps we should just note my method as a shortcut for people with no interest in beyond Docker.

Author: Adding in a wrapper to cover both.

```
tpaexec ping demo
```

to ping all the SSH connectable hosts in the deployment: You will get output that looks something like:

```
$ tpaexec ping demo 
unfair | SUCCESS => {
    "changed": false,
    "ping": "pong"
}
uptake | SUCCESS => {
    "changed": false,
    "ping": "pong"
}
quondam | SUCCESS => {
    "changed": false,
    "ping": "pong"
}
uptight | SUCCESS => {
    "changed": false,
    "ping": "pong"
}
```

> Select one of the nodes which has responded with `SUCCESS`. We shall use `uptake` foir this example.

Reviewer: ```suggestion
Select one of the nodes which responded with `SUCCESS`. We shall use `uptake` for this example.
```

Author: ooops

Change current directory to the created configuration directory. For example, our configuration is called demo, so we go to that directory. In there, we run `ssh -F ssh_config ourhostname` to connect.

```
cd demo
ssh -F ssh_config uptake
Last login: Wed Sep  6 10:08:01 2023 from 172.17.0.1
[root@uptake ~]# 
```

We can now change user to the `postgres` user using `sudo -iu postgres`. As `postgres` we can run `psql`. TPA has already configured that user with a `.pgpass` file so there's no need to present a password.

```
[root@uptake ~]# 
postgres@uptake:~ $ psql
psql (15.4)
Type "help" for help.

postgres=# 
```

And we are connected to our database.

You can connect from the host system without SSHing into one of the containers. Obtain the IP address of the host you want to connect to from the `ssh_config` file.

```
$ grep "^ *Host" demo/ssh_config 
Host *
Host uptight
    HostName 172.17.0.9
Host unfair
    HostName 172.17.0.4
Host quondam
    HostName 172.17.0.10
Host uptake
    HostName 172.17.0.11
```

> We are gping to connect to uptake, so the IP address is 172.17.0.11. 

Reviewer: ```suggestion
We are going to connect to uptake, so the IP address is 172.17.0.11. 
```

Author: Ah my spell checker has died. Two things to fix.
